samples/client/petstore/rust/reqwest/fileResponseTest/src/apis/default_api.rs

```diff
@@ -73,9 +63,7 @@ impl DefaultApi for DefaultApiClient {
         let content = resp.text()?;
 
         if status.is_success() {
-            let entity: Option<FileresponsetestSuccess> = serde_json::from_str(&content).ok();
-            let result = ResponseContent { status, content, entity };
-            Ok(result)
+            serde_json::from_str(&content).map_err(Error::from)
         } else {
             let entity: Option<FileresponsetestError> = serde_json::from_str(&content).ok();
```

Comment:
in light of [this](https://github.com/OpenAPITools/openapi-generator/pull/6673#issuecomment-645357769), I don't think it's possible for `serde_json::from_str()` to fail because you've already covered `serde_json::Value`

`ResponseContent` should become:

```rust
struct ResponseContent {
    entity: T,
    status: StatusCode,
}
```

Reply:
As you pointed out, the call to `serde_json::from_str()` won't fail due to unexpected Types as we have covered `serde_json::Value` - but it might fail when `content` does not contain valid JSON.

Reply:
hmm, then we should use the `response.json()` method.

Reply:
I'm afraid this will not help, as in case of invalid JSON [`reqwest::Response.json()`](https://docs.rs/reqwest/0.10.6/reqwest/struct.Response.html#method.json) will also fail:
> `pub async fn json<T: DeserializeOwned>(self) -> Result<T>`
> Try to deserialize the response body as JSON.
> [...]
> Errors
> This method fails whenever the response body is not in JSON format or it cannot be properly deserialized to target type `T`. For more details please see `serde_json::from_reader`.

Reply:
yes but then we already have an error for reqwest defined.

```rust
#[derive(Debug)]
pub enum Error<T> {
    Reqwest(reqwest::Error),
    Serde(serde_json::Error),
    Io(std::io::Error),
    ResponseError(ResponseContent<T>),
}
```

so code becomes

```rust
let status = resp.status();
let content = resp.text().await?;

if status.is_success() {
    // would recommend using derive_more::From for the error type
    Ok(response.json().await.map_err(Error::Reqwest)?)
} else {
        let entity: CloseError = response.json().await.map_err(Error::Reqwest)?;
        let error = ResponseContent { status, entity };
        Err(Error::ResponseError(error))
}
```

Reply:
Good idea! That way we don't have to deal with `serde_json::error::Error` explicitily. As a downside, the raw content would no longer be available in the result. There was probably a reason for adding this in the first place, maybe @bcourtine can shed some light on this.

As for your remark on `derive_more::From`:
`impl <T> From<reqwest::Error> for Error<T>` is already present in `api_mod.mustache`. The `map_err()` should therefore not be necessary. Or do I overlook something here?

Reply:
> As a downside, the raw content would no longer be available in the result.

since the response content is malformed json, the only thing useful to do with it is log to stdout/err or how would you try to handle it?

Reply:
> 
> 
> > As a downside, the raw content would no longer be available in the result.
> 
> since the response content is malformed JSON, the only thing useful to do with it is log to stdout/err or how would you try to handle it?

I meant in case of success: The current implementation always provides the raw content through `ResponseContent.content` (which requires using `response.text()` and `serde_json::from_str()`). Your suggestion would remove this. Sorry for the confusion!

Reply:
hmm, I'm not sure how this happens because the specification defines the response content-type.

```json
"produces": [
    "application/json"
]
```

and so if the success content isn't valid JSON shouldn't that be an error?

Reply:
My remark about dropping the raw content wasn't related to error handling - sorry for being too vague here. I was referring to your suggestion of changing the return value from
```rust
let status = resp.status();
let content = resp.text().await?;
if status.is_success() {
    let entity: Option<UsersSetRegularSuccess> = serde_json::from_str(&content).ok();
    let result = ResponseContent { status, content, entity };
    Ok(result)
} ...
```
to this:
```rust
let status = resp.status();
if status.is_success() {
   Ok(resp.json().await?) // .map_err(Error::Reqwest) should not be necessary
} ...
```
With this, `status` and `content` will no longer be provided to the API user in case of success. I do not know why this might be necessary or useful - however, it is provided by the current version of the code for some reason.

Reply:
> My remark about dropping the raw content wasn't related to error handling - sorry for being too vague here. I was referring to your suggestion of changing the return value from
> ```rust
> let status = resp.status();
> let content = resp.text().await?;
> if status.is_success() {
>     let entity: Option<UsersSetRegularSuccess> = serde_json::from_str(&content).ok();
>     let result = ResponseContent { status, content, entity };
>     Ok(result)
> } ...
> ```
> to this:
> ```rust
> let status = resp.status();
> if status.is_success() {
>    Ok(resp.json().await?) // .map_err(Error::Reqwest) should not be necessary
> } ...
> ```
> With this, `status` and `content` will no longer be provided to the API user in case of success. I do not know why this might be necessary or useful - however, it is provided by the current version of the code for some reason.

This pr removes use of `ResponseContent` in the success case, actually.

What I'm advocating for is removing the content field in `ResponseContent`

Reply:
> This pr removes use of ResponseContent in the success case, actually.

ResponseContent is not dropped: it becomes optional. If you use `supportMultipleResponses=true`, you will still get a `ResponseContent` in success case.

I totally dropped ResponseContent with `supportMultipleResponses=false` because the PR aimed at providing the simplest possible signature, like in 4.x.x version.

`content` field exists mostly for error case: it allows to debug an error (malformed JSON which cannot be converted to entity). I let this field in success case to avoid multiplicating structs.

First, I will simplify enum values dropping `UnknowList` (included in `UnknownValue`).

For the rest, I am a little lost in this debate… do you have a consensus on the best possible solution (for both `supportMultipleResponses` values)?

Reply:
hopefully, @bcourtine's comment clarifies things @HenningHolmDE

to clarify further:

> What I'm advocating for is removing the `content` field in `ResponseContent`

because:

> since the response content is malformed JSON, the only thing useful to do with it is: log to stdout/err

Reply:
> since the response content is malformed JSON, the only thing useful to do with it is: log to stdout/err

It may or may not be malformed. We've seen issues in which a different type of object was returned (e.g. expecting array of Dog but get a single Dog, expecting a Cat but get a Dog due to invalid spec, expecting JSON but got XML as XML is the default when `accept` is not specified in the request header in that particular case) so my recommendation is still to include it in the Error response object to make developer's life easier in troubleshooting what's wrong.

Reply:
I was under the impression that logging the reqwest Error had the response text, but after testing i just confirmed it doesn't so this is fine.

> include it in the Error response object to make developer's life easier in troubleshooting what's wrong.

I think that's the last thing, we can finally get this merged? :confetti_ball: :partying_face:

Reply:
> I think that's the last thing, we can finally get this merged? :confetti_ball: :partying_face:

No objections from my side. :+1:

Reply:
👍 👍 👍